### PR TITLE
current progress

### PR DESCRIPTION
**Pull request by Nuclearl**
https://drive.google.com/drive/folders/0B3jvg6jQR0D9QlV3RjducDEta3M?resourcekey=0-adUUXDhvH0f4BJ_sh7r3UQ

### CODE REVIEW

File: app/models/book.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  validates :height, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :width, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :depth, presence: true, numericality: { greater_than_or_equal_to: 0 }
+
+  NUMBER_PHOTOS = 4
+
+  private
+
+  def validate_number_images
+    errors.add(:images, 'is too many') if images.size > NUMBER_PHOTOS
```
Comment (WoLkErSs) [inline app/models/book.rb:28]: I18n

File: app/models/book.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
+class Book < ApplicationRecord
+  validate :validate_number_images
+
+  has_many :authors_books, dependent: :destroy
+  has_many :authors, through: :authors_books
+
+  has_many :books_sections, dependent: :destroy
```
Comment (WoLkErSs) [inline app/models/book.rb:7]: rename to categories

File: app/models/book.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  validates :title, presence: true
+  validates :price, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :description, presence: true
+  validates :published_at, presence: true, numericality: { only_integer: true, greater_than_or_equal_to: 1 }
+  validates :materials, presence: true
+  validates :height, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :width, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :depth, presence: true, numericality: { greater_than_or_equal_to: 0 }
+
+  NUMBER_PHOTOS = 4
```
Comment (WoLkErSs) [inline app/models/book.rb:23]: move under class name

File: app/models/section.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class Section < ApplicationRecord
```
Comment (WoLkErSs) [inline app/models/section.rb:1]: rename to category

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
+class User < ApplicationRecord
```
Comment (WoLkErSs) [inline app/models/user.rb:1]: use VCR to test it

File: app/services/books_services/filters.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      'Popular first' => '',
+      'Price: Low to hight' => 'price ASC',
+      'Price: Hight to low' => 'price DESC'
+    }.freeze
+
+    DEFAULT_FILTER = FILTERS.keys.first
+
+    def initialize(params)
+      @params = params
+      super()
```
Comment (WoLkErSs) [inline app/services/books_services/filters.rb:16]: remove

File: app/services/books_services/filters.rb [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      [sort_books_by_filters(books, filter), section, filter]
+    end
+
+    private
+
+    def sort_books_by_filters(books, filter)
+      if filter && FILTERS.include?(filter)
+        books.order(FILTERS[filter])
+      else
+        Book.all
```
Comment (WoLkErSs) [inline app/services/books_services/filters.rb:32]: have to returne books
add specs on this case

File: app/services/books_services/filters.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    def initialize(params)
+      @params = params
+      super()
+    end
+
+    def call
+      section = Section.find(params[:section]) if params[:section]
+      filter = params[:filter] ? params[:filter] : DEFAULT_FILTER
+      books = receive_books_from_sections(section)
```
Comment (WoLkErSs) [inline app/services/books_services/filters.rb:22]: do not use words like `receive` or `get` in method name

File: spec/controllers/address_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
+require 'rails_helper'
```
Comment (WoLkErSs) [inline spec/controllers/address_controller_spec.rb:1]: remove

**Review by WoLkErSs — CHANGES_REQUESTED** [review]
Move all text to I18n

**Comment by Nuclearl** [comment]
https://drive.google.com/drive/folders/0B3jvg6jQR0D9QlV3RjducDEta3M?resourcekey=0-adUUXDhvH0f4BJ_sh7r3UQ

File: config/environments/production.rb [hunk 10]
```diff
@@ -1,120 +1,61 @@
...
-  # Code is not reloaded between requests.
+  config.action_mailer.delivery_method = :smtp
+  config.action_mailer.default_url_options = { host: 'blooming-river-38869.herokuapp.com' }
+  config.action_mailer.perform_deliveries = true
+  config.action_mailer.raise_delivery_errors = true
+  config.action_mailer.default :charset => "utf-8"
+  config.action_mailer.smtp_settings = {
+    :address => "smtp.gmail.com",
+    :port => 587,
+    :domain => 'blooming-river-38869.herokuapp.com',
```
Comment (WoLkErSs) [inline config/environments/production.rb:13]: move to credentials

File: config/environments/test.rb [hunk 11]
```diff
@@ -14,9 +22,10 @@
   # Do not eager load code on boot. This avoids loading your whole application
   # just for the purpose of running a single test. If you are using a tool that
   # preloads Rails for running tests, you may have to set it to true.
-  config.eager_load = false
+  config.eager_load = true
```
Comment (WoLkErSs) [inline config/environments/test.rb:25]: read about eager_load, please

File: spec/controllers/address_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+      it { is_expected.to redirect_to(settings_path) }
+    end
+
+    context 'when address form is invalid' do
+      before do
+        patch :update, params: { id: address.id, address_form: address.attributes.merge({ 'first_name' => nil }) },
+                       format: :js
+      end
+
+      it { is_expected.to respond_with(200) }
```
Comment (WoLkErSs) [inline spec/controllers/address_controller_spec.rb:37]: check current page

File: spec/controllers/books_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  describe 'GET index' do
+    let(:books) { Book.all }
+
+    let(:category) { 1 }
+    let(:filter) { 'random' }
+
+    before do
+      my_services = instance_double('Filters')
```
Comment (WoLkErSs) [inline spec/controllers/books_controller_spec.rb:18]: move to let

Comment (WoLkErSs) [reply spec/controllers/books_controller_spec.rb:18]: use Constant not string

File: spec/controllers/review_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        post :create, params: { review_form: review_params }
+      end
+
+      it do
+        expect(controller).to permit(:score, :title, :review, :book_id, :user_id)
+          .for(:create, params: { review_form: review_params })
+          .on(:review_form)
+      end
+
+      it { expect(Review.all.size).to eq 1 }
```
Comment (WoLkErSs) [inline spec/controllers/review_controller_spec.rb:25]: use change by

File: spec/controllers/settings_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+      get :index
+    end
+
+    it { expect(assigns[:billing_address]).to be_instance_of Address }
+    it { expect(assigns[:shipping_address]).to be_instance_of Address }
+
+    it { expect(user.addresses).to include assigns[:billing_address] }
+    it { expect(user.addresses).to include assigns[:shipping_address] }
+
+    it { expect(user.addresses.size).to eq 2 }
```
Comment (WoLkErSs) [inline spec/controllers/settings_controller_spec.rb:21]: you have to check exactly type of addresses

File: spec/controllers/settings_controller_spec.rb [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+
+  describe 'DELETE destroy' do
+    before do
+      delete :destroy
+    end
+
+    it { expect(flash[:success]).to eq I18n.t('notification.delete_account') }
+    it { is_expected.to respond_with(302) }
+    it { is_expected.to redirect_to(root_path) }
+    it { expect(user.destroy).to be_truthy }
```
Comment (WoLkErSs) [inline spec/controllers/settings_controller_spec.rb:99]: validate that amount changed by 1

File: spec/controllers/welcome_controller_spec.rb [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  describe 'GET index' do
+    before do
+      FactoryBot.create_list(:book, 5)
+      get :index
+    end
+
+    it { expect(assigns[:latest_books]).to be_instance_of Array }
+    it { expect(assigns[:latest_books].size).to eq 3 }
```
Comment (WoLkErSs) [inline spec/controllers/welcome_controller_spec.rb:15]: use constant

File: spec/controllers/welcome_controller_spec.rb [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+    FactoryBot.create_list(:category, 3)
+  end
+
+  describe 'GET index' do
+    before do
+      FactoryBot.create_list(:book, 5)
+      get :index
+    end
+
+    it { expect(assigns[:latest_books]).to be_instance_of Array }
```
Comment (WoLkErSs) [inline spec/controllers/welcome_controller_spec.rb:14]: check what inside array

File: spec/factories/books.rb [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+    materials { FFaker::Color.name }
+
+    trait :with_category do
+      after(:create) do |book|
+        book.categories << FactoryBot.create(:category)
+      end
+    end
+
+    after(:build) do |book|
+      4.times do
```
Comment (WoLkErSs) [inline spec/factories/books.rb:19]: do not save 4 photos, 1 will be enough

File: spec/features/authentication/forgot_password_spec.rb [hunk 20]
```diff
@@ -0,0 +1,37 @@
+describe 'Forgot password' do
+  let(:user) { FactoryBot.create(:user, email: 'test@test.com') }
```
Comment (WoLkErSs) [inline spec/features/authentication/forgot_password_spec.rb:2]: use ffaker

File: spec/features/authentication/restore_password_spec.rb [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+        fill_in 'user[password]', with: new_password
+        fill_in 'user[password_confirmation]', with: new_password
+      end
+      click_button I18n.t('button.change')
+    end
+
+    it { expect(page).to have_content I18n.t('devise.passwords.updated') }
+  end
+
+  context 'when change password if url is incorrect' do
```
Comment (WoLkErSs) [inline spec/features/authentication/restore_password_spec.rb:25]: rename

Comment (WoLkErSs) [inline spec/features/authentication/restore_password_spec.rb:25]: add test when invalid token

File: spec/features/authentication/restore_password_spec.rb [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+
+  context 'when reset password' do
+    before do
+      visit new_user_password_path
+      within '#new_user' do
+        fill_in 'user[email]', with: user.email
+        click_button I18n.t('button.email_instructions')
+      end
+
+      open_email(user.email, with_subject: 'Reset password instructions')
```
Comment (WoLkErSs) [inline spec/features/authentication/restore_password_spec.rb:13]: use i18n

File: spec/features/authentication/restore_password_spec.rb [hunk 23]
```diff
@@ -0,0 +1,33 @@
+describe 'Forgot password' do
```
Comment (WoLkErSs) [inline spec/features/authentication/restore_password_spec.rb:1]: add tests for all cases

**Review by WoLkErSs — CHANGES_REQUESTED** [review]

File: spec/services/books_services/filters_spec.rb [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+      it 'returns books from this category' do
+        expect(described_class.new(params).call).to eq [category.books.order(default_order), category, default_filter]
+      end
+    end
+
+    context 'when filter is defined' do
+      let(:params) { { filter: second_filter } }
+
+      it 'returns books this category' do
+        expect(described_class.new(params).call).to eq [Book.all.order(second_order), nil, second_filter]
```
Comment (WoLkErSs) [inline spec/services/books_services/filters_spec.rb:29]: move to subject

File: spec/services/books_services/filters_spec.rb [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+  let(:default_order) { 'published_at ASC' }
+  let(:default_filter) { 'asc' }
+  let(:second_order) { 'published_at DESC' }
+  let(:second_filter) { 'desc' }
+
+  before do
+    stub_const('BooksServices::Filters::FILTERS',
+               { default_filter => default_order, second_filter => second_order })
+    stub_const('BooksServices::Filters::DEFAULT_FILTER', default_filter)
+    FactoryBot.create_list(:book, 5, categories: [category])
```
Comment (WoLkErSs) [inline spec/services/books_services/filters_spec.rb:12]: create less entities as it posible

**Review by WoLkErSs — APPROVED** [review]
Fix please comment and merge

**Pull request merged**